Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,740 @@
+@@ -32,14 +32,19 @@ public final class PlayerMobDistanceMap {
+         return this.playerMap.getOrDefault(ChunkCoordIntPair.pair(chunkX, chunkZ), EMPTY_SET);
+     }
+ 
+-    public void update(final List<EntityPlayer> currentPlayers, final int newViewDistance) {
++    public void update(final List<EntityPlayer> currentPlayers) { //Paper ticket level pp-view-dis
+         AsyncCatcher.catchOp("Distance map update");
+         final ObjectLinkedOpenHashSet<EntityPlayer> gone = new ObjectLinkedOpenHashSet<>(this.players.keySet());
+ 
+-        final int oldViewDistance = this.viewDistance;
+-        this.viewDistance = newViewDistance;
++        //Paper start ticket level pp-view-dist
++        //final int oldViewDistance = this.viewDistance;
++        //this.viewDistance = newViewDistance;
+ 
+         for (final EntityPlayer player : currentPlayers) {
++            player.oldMobViewDistance = player.mobViewDistance;
++            player.mobViewDistance = player.getViewDistance();
++            //Paper end
++
+             if (player.isSpectator() || !player.affectsSpawning) {
+                 continue; // will be left in 'gone' (or not added at all)
+             }
+@@ -50,9 +55,10 @@ public final class PlayerMobDistanceMap {
+             final SectionPosition oldPosition = this.players.put(player, newPosition);
+ 
+             if (oldPosition == null) {
+-                this.addNewPlayer(player, newPosition, newViewDistance);
++                this.addNewPlayer(player, newPosition, player.mobViewDistance); //Paper ticket level pp-view-dist
+             } else {
+-                this.updatePlayer(player, oldPosition, newPosition, oldViewDistance, newViewDistance);
++                this.updatePlayer(player, oldPosition, newPosition, player.oldMobViewDistance, player.mobViewDistance); //Paper ticket level pp-view-dist
++
+             }
+             //this.validatePlayer(player, newViewDistance); // debug only
+         }
+@@ -60,7 +66,7 @@ public final class PlayerMobDistanceMap {
+         for (final EntityPlayer player : gone) {
+             final SectionPosition oldPosition = this.players.remove(player);
+             if (oldPosition != null) {
+-                this.removePlayer(player, oldPosition, oldViewDistance);
++                this.removePlayer(player, oldPosition, player.oldMobViewDistance); //Paper ticket level pp-view-dist
+             }
+         }
+     }
+@@ -115,7 +121,12 @@ public final class PlayerMobDistanceMap {
+ 
+     private void removePlayerFrom(final EntityPlayer player, final int chunkX, final int chunkZ) {
+         this.playerMap.compute(ChunkCoordIntPair.pair(chunkX, chunkZ), (final Long keyInMap, final PooledHashSets.PooledObjectLinkedOpenHashSet<EntityPlayer> players) -> {
+-            return PlayerMobDistanceMap.this.pooledHashSets.findMapWithout(players, player); // rets null instead of an empty map
++            //Paper start ticket level pp-view-dist
```

Comment:
revert this, there is no reason for this change

Reply:
There is reason, one of the testers found (although in a very rare case) that with rapid async teleportation (from paperlib) with a view distance changes in between this will occur:

https://pastebin.com/hk0uG4zQ

It is better to have this anyways in general.

Reply:
Look at the change you made to the function named `removePlayerFrom`: You will ADD a player if they're not in the area. This is a dumb fix to whatever issue you have, and that issue is most likely YOUR fault. Debug the issue correctly, and you will find the solution is not whatever this is.

Reply:
I'll double check it

Reply:
It's been removed, I'll test normal global view distance changes against this and see what happens.

Reply:
This error still occurs under those conditions, so this is not per-player specific. It may be a very specific rare case with the plugin handling async teleportation wrong.

Reply:
Just tested this many more times, this issue is independent of per player view distance.

Reply:
well if you can roll out a case where it's the mob spawn map's fault then lay out the issue, and either I can PR a fix or you can include one in this PR
In either case the above fix is still inappropriate